### PR TITLE
Link to Tags using sluggedName instead of the name itself.

### LINKED ISSUE #119: Invalid tag format should throw error
Creating an invalid tag when posting a question should throw error instead of directing to a new window.

invalid tag

### LINKED ISSUE #120: Tags with dots
Tags with dots should be allowed as valid format. We have a use case where version numbers are used for tags. eg: v3.2, v5.10

### PR DESCRIPTION
**Pull request by MatejBalantic**
* this allows tag names to contain any character, including spaces etc, thus validation of the name is now removed
* slugged name is automatically generated in the same way as for the questions
* the change is backwards-compatible as all the names are transferred to suggedName upon migration

Notice: this passes unit tests, however will not build on Travis until https://github.com/caelum/mamute/pull/134 is merged

Closes #120
Closes #119

### CODE REVIEW

File: src/main/java/org/mamute/controllers/ListController.java [hunk 1]
```diff
@@ -159,9 +159,9 @@ public void unanswered(Integer p) {
 	}
 	
 	@Get
-	public void withTag(String tagName, Integer p, boolean semRespostas) {
+	public void withTag(String tagUriName, Integer p, boolean semRespostas) {
```
Comment (leocwolter) [inline src/main/java/org/mamute/controllers/ListController.java]: that variable `semRespostas` should be `withoutAnswers`. It's in portuguese ):

File: src/main/java/org/mamute/controllers/RankingController.java [hunk 2]
```diff
@@ -34,8 +34,8 @@ public void rank(Integer p) {
 	}
 	
 	@Get
-	public void tagRank(String tagName){
-		Tag tag = tags.findByName(tagName);
+	public void tagRank(String tagUriName){
```
Comment (leocwolter) [inline src/main/java/org/mamute/controllers/RankingController.java]: Better not to change controllers methods parameter's names. They are used at [`routes.properties`](https://github.com/caelum/mamute/blob/master/src/main/resources/routes.properties) files and someone could overwrite those.

File: src/test/java/org/mamute/dao/TagDAOTest.java [hunk 3]
```diff
@@ -141,6 +150,25 @@ public void should_not_save_repeated_tags() throws Exception {
...
+        Tag cevapciciWithoutSpecialChar = new Tag("cevapcici", "", leo);
+        tags.saveIfDoesntExists(cevapciciWithSpecialChar);
+        tags.saveIfDoesntExists(cevapciciWithoutSpecialChar);
+
+        assertEquals(tags.all().size(), originalSize+1);
+    }
+
+    @Test
+    public void should_define_correct_slugged_url()
+    {
```
Comment (leocwolter) [inline src/test/java/org/mamute/dao/TagDAOTest.java]: we open curly braces at the same line as the method signature :)

**Comment by leocwolter** [comment]
Since we support the configurations below at mamute instances, the user can use a dash to split tags ):
```
# Regular expression to split tags inside Java
tags.splitter.regex = \\,
# Char used to split tags inside the JSP forms. It must matchs with the tags.splitter.regex property
tags.splitter.char = ,
# Regex of allowed characters for tag names
tags.sanitizer.regex = [a-zA-Z0-9-]
```
I think we should use them when generating the slugged tags(maybe the Normalizer should be customizable too)

**Comment by MatejBalantic** [comment]
I noticed that but I didn't see any case where this would matter. Is there
any? Can you provide an example of string that could cause issues?


On Thursday, January 15, 2015, Leonardo Cesar Wolter <
notifications@github.com> wrote:

> Since we support the configurations below at mamute instances, the user
> can use a dash to split tags ):
>
> # Regular expression to split tags inside Java
> tags.splitter.regex = \\,
> # Char used to split tags inside the JSP forms. It must matchs with the tags.splitter.regex property
> tags.splitter.char = ,
> # Regex of allowed characters for tag names
> tags.sanitizer.regex = [a-zA-Z0-9-]
>
> I think we should use them when generating the slugged tags(maybe the
> Normalizer should be customizable too)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/caelum/mamute/pull/135#issuecomment-70033562>.
>


-- 

*Matej Balantič*

T +49 171 2976495
W http://about.me/Matej.Balantic

File: src/main/java/org/mamute/controllers/RankingController.java (hunk 2 above)
Comment (MatejBalantic) [inline src/main/java/org/mamute/controllers/RankingController.java]: Thanks, good point. Will revert the routes back

File: src/test/java/org/mamute/dao/TagDAOTest.java (hunk 3 above)
Comment (MatejBalantic) [inline src/test/java/org/mamute/dao/TagDAOTest.java]: My apologies, I'm coming from Objective-C :) Will fix it of course

File: src/main/java/org/mamute/controllers/ListController.java (hunk 1 above)
Comment (MatejBalantic) [inline src/main/java/org/mamute/controllers/ListController.java]: I noticed, but resisted an urge to fix it :) I'll change it with other things

**Comment by leocwolter** [comment]
I could use dash to split and allow spaces on my tags: java ee-jpa
The final result would be java-ee and jpa. If I decide to update those tags, it would be shown as java-ee-jpa :)

**Comment by MatejBalantic** [comment]
That shouldn't be true as only the sluggedName would contain the dash, not
the name. So you would still end up with
java ee
jpa

after insert which would work fine when updating. Maybe I just don't
understand, can you write a unit test which would cause the issue you are
talking about?

On Thursday, January 15, 2015, Leonardo Cesar Wolter <
notifications@github.com> wrote:

> I could use dash to split and allow spaces on my tags: java ee-jpa
> The final result would be java-ee and jpa. If I decide to update those
> tags, it would be shown as java-ee-jpa :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/caelum/mamute/pull/135#issuecomment-70085693>.
>


-- 

*Matej Balantič*

T +49 171 2976495
W http://about.me/Matej.Balantic

File: src/main/java/org/mamute/model/Tag.java [hunk 4]
```diff
@@ -28,7 +30,11 @@
 	@Column(unique = true, nullable = false)
 	@NotEmpty
 	private String name;
-	
+
+    @Type(type = "text")
+    @NotEmpty
+    private String sluggedName;
```
Comment (leocwolter) [inline src/main/java/org/mamute/model/Tag.java]: that one should have unique and not null constraint  in it too, just in case :)

**Comment by leocwolter** [comment]
Oh, we are keeping the name and the sluggedName, got it. Sorry, I thought you wanted to normalize the name of the tags :)

File: src/main/java/org/mamute/model/Tag.java [hunk 5]
```diff
@@ -48,7 +54,8 @@ public Tag() {
 	
 	public Tag(String name, String description, User author) {
 		this.name = name.toLowerCase();
-		this.description = description;
+        this.sluggedName = toSlug(name);
+        this.description = description;
```
Comment (leocwolter) [inline src/main/java/org/mamute/model/Tag.java]: We use tab for indentation(with size of 4 blank spaces), can you configure your editor for that?

File: src/main/java/org/mamute/model/Tag.java [hunk 6]
```diff
@@ -57,12 +64,7 @@ public String getName() {
 	}
 	
 	public String getUriName() {
-		try {
-			return URLEncoder.encode(name, "utf-8");
-		} catch (UnsupportedEncodingException e) {
-			e.printStackTrace();
-		}
-		return null;
+		return this.sluggedName;
```
Comment (leocwolter) [inline src/main/java/org/mamute/model/Tag.java]: Could you test tags like `c++` and `c#` please? The slugged name cannot be just `c` since it would conflict

File: src/main/resources/mamute.properties [hunk 7]
```diff
@@ -14,8 +14,6 @@ vraptor.errorcontrol.date.joda.pattern = dd/MM/yyyy
 tags.splitter.regex = \\,
 # Char used to split tags inside the JSP forms. It must matchs with the tags.splitter.regex property
 tags.splitter.char = ,
-# Regex of allowed characters for tag names
-tags.sanitizer.regex = [a-zA-Z0-9-]
```
Comment (leocwolter) [inline src/main/resources/mamute.properties]: Sorry, I didn't get it, why did you remove this?

File: src/test/java/org/mamute/model/TagTest.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    {
+        assertEquals(subject.getDescription(), description);
+    }
+
+    @Test
+    public void testGetUriName()
+    {
+        assertEquals(subject.getUriName(), "tag-with-specialchars");
+    }
+}
```
Comment (leocwolter) [inline src/test/java/org/mamute/model/TagTest.java]: same here {}  :)

File: src/main/resources/mamute.properties (hunk 7 above)
Comment (MatejBalantic) [inline src/main/resources/mamute.properties]: I removed it as we are no longer validating the tag names (not needed due
to slugs). Will have to rethink the approach to get c++ and c# working
together though. Thanks for pointing this out.

On Thursday, January 15, 2015, Leonardo Cesar Wolter <
notifications@github.com> wrote:

> In src/main/resources/mamute.properties
> <https://github.com/caelum/mamute/pull/135#discussion-diff-23029896>:
>
> > @@ -14,8 +14,6 @@ vraptor.errorcontrol.date.joda.pattern = dd/MM/yyyy
> >  tags.splitter.regex = \\,
> >  # Char used to split tags inside the JSP forms. It must matchs with the tags.splitter.regex property
> >  tags.splitter.char = ,
> > -# Regex of allowed characters for tag names
> > -tags.sanitizer.regex = [a-zA-Z0-9-]
>
> Sorry, I didn't get it, why did you remove this?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/caelum/mamute/pull/135/files#r23029896>.
>


-- 

*Matej Balantič*

T +49 171 2976495
W http://about.me/Matej.Balantic

Comment (leocwolter) [inline src/main/resources/mamute.properties]: We could just use the Normalizer to url-encode those things, I guess it makes sense on every case we normalize a string, but maybe we should rename Normalizer to make this more explicit

Comment (MatejBalantic) [inline src/main/resources/mamute.properties]: We need to transliterate characters like čžš to czs, and only than we could url-encode non-transliteratable characters like # and +. I'll prepare something and commit it here.

File: src/test/java/org/mamute/model/TagTest.java (hunk 8 above)
Comment (MatejBalantic) [inline src/test/java/org/mamute/model/TagTest.java]: I'll check them all out, sorry about that again.

File: src/main/java/org/mamute/model/Tag.java (hunk 5 above)
Comment (MatejBalantic) [inline src/main/java/org/mamute/model/Tag.java]: Of course. Should this project maybe have a wiki with coding standards and editor settings explained? I imagine most of them are industry standard in Java world, but not everybody are from Java world (khm, people like me:)

Comment (csokol) [inline src/main/java/org/mamute/model/Tag.java]: I think we can create a "Contributing" section in our readme or a question at meta.mamute.org for that...

Comment (leocwolter) [inline src/main/java/org/mamute/model/Tag.java]: Yeah, or maybe we can add the eclipse / intellij code style file at git

**Comment by MatejBalantic** [comment]
Hi, so I amended the pull requests:
* addressed the review comments regarding indentation and brackets, where I noticed it at least,
* slugging for tags is now done by normalizing alphabetic characters and then url encoding special characters. Also added tests for C# and c++

File: src/main/java/org/mamute/controllers/TagPageController.java [hunk 9]
```diff
@@ -47,24 +47,24 @@ public void editTagPage(String tagName, MarkedText about){
...
+		Tag tag = tags.findBySluggedName(tagName);
 		TagPage tagPage = new TagPage(tag, about);
 		if(!validator.validate(tagPage)){
-			validator.onErrorRedirectTo(TagPageController.class).tagPageForm(tagPage.getTagName());
+			validator.onErrorRedirectTo(TagPageController.class).tagPageForm(tagPage.getTag().getUriName());
 			return;
 		}
 		tagPages.save(tagPage);
-		result.redirectTo(this).showTagPage(tagPage.getTagName()); 
+		result.redirectTo(this).showTagPage(tagPage.getTag().getUriName());
```
Comment (leocwolter) [inline src/main/java/org/mamute/controllers/TagPageController.java]: Since we will always need `tagPage.getTag().getUriName()` to get the uri, we could encapsulate this inside tagPage. Something like `tagPage.getTagUriName()` or a better name :P

File: src/main/java/org/mamute/dao/TagDAO.java [hunk 10]
```diff
@@ -88,11 +91,13 @@ public boolean hasAbout(Tag tag) {
 	}
 
 	public Tag saveIfDoesntExists(Tag newTag) {
-		Tag existingTag = findByName(newTag.getName());
+		Tag existingTag = findBySluggedName(newTag.getUriName());
```
Comment (leocwolter) [inline src/main/java/org/mamute/dao/TagDAO.java]: Better to use the same name for both: "sluggedName" or "uriName"

File: src/main/java/org/mamute/infra/NormalizerBrutal.java [hunk 11]
```diff
@@ -1,18 +1,34 @@
...
 import java.util.Locale;
 import java.util.regex.Pattern;
 
 public class NormalizerBrutal {
-	public static String toSlug(String input) {
+	public static String toSlug(String input, boolean urlEncodeNonNormalizableCharacter)  {
 		if(input == null || input.isEmpty()) return "";
-		Pattern NONLATIN = Pattern.compile("[^\\w-]");
+
+		String nonLatinPattern = urlEncodeNonNormalizableCharacter ? "[^\\P{M}]" : "[^\\w-]";
```
Comment (leocwolter) [inline src/main/java/org/mamute/infra/NormalizerBrutal.java]: Sorry, but why was it necessary? Didn't get it

File: src/main/java/org/mamute/infra/NormalizerBrutal.java [hunk 12]
```diff
@@ -1,18 +1,34 @@
...
+		Pattern NONLATIN = Pattern.compile(nonLatinPattern);
 		Pattern WHITESPACE = Pattern.compile("[\\s]");
 		String nowhitespace = WHITESPACE.matcher(input).replaceAll("-");
 		String normalized = Normalizer.normalize(nowhitespace, Form.NFD);
 		String slug = NONLATIN.matcher(normalized).replaceAll("");
+
+		if (urlEncodeNonNormalizableCharacter) {
+			try {
+				slug = URLEncoder.encode(slug, "UTF-8");
+			} catch (UnsupportedEncodingException e) {	}
```
Comment (leocwolter) [inline src/main/java/org/mamute/infra/NormalizerBrutal.java]: `throw new IllegalArgumentException(e)` just in case :)

File: src/main/java/org/mamute/model/Tag.java [hunk 13]
```diff
@@ -24,45 +26,46 @@
...
+
 	@Column(unique = true, nullable = false)
 	@NotEmpty
 	private String name;
-	
+
+    @Column(unique = true, nullable = false)
+    @Type(type = "text")
+    @NotEmpty
+    private String sluggedName;
```
Comment (leocwolter) [inline src/main/java/org/mamute/model/Tag.java]: identation :(

File: src/main/java/org/mamute/model/Tag.java [hunk 14]
```diff
@@ -24,45 +26,46 @@
...
 	public Tag() {
 		this("", "", null);
 	}
-	
+
 	public Tag(String name, String description, User author) {
 		this.name = name.toLowerCase();
-		this.description = description;
+        this.sluggedName = toSlug(name, true);
+        this.description = description;
```
Comment (leocwolter) [inline src/main/java/org/mamute/model/Tag.java]: identation :(

File: src/test/java/org/mamute/dao/TagDAOTest.java [hunk 15]
```diff
@@ -141,12 +172,28 @@ public void should_not_save_repeated_tags() throws Exception {
...
+
+        assertEquals(tags.all().size(), originalSize+1);
+    }
+
+    @Test
+    public void should_define_correct_slugged_url() {
+        assertNotNull(tags.findBySluggedName("java"));
+        assertNotNull(tags.findBySluggedName("ruby"));
+        assertNotNull(tags.findBySluggedName("c%2b%2b"));
+    }
```
Comment (leocwolter) [inline src/test/java/org/mamute/dao/TagDAOTest.java]: better to assert if the returned tag is actually `c++`

File: src/test/java/org/mamute/dao/TagDAOTest.java [hunk 16]
```diff
@@ -88,17 +96,32 @@ public void should_get_main_tags_of_the_provided_user() throws Exception {
...
+    public void should_get_existing_tags_with_different_name_same_slug() throws Exception {
+        List<Tag> found = tags.findAllDistinct(asList("cemaz"));
+        assertFalse(found.isEmpty());
+    }
+
+	@Test
+	public void should_get_existing_tags_with_different_name_different_slug() throws Exception {
+		List<Tag> found = tags.findAllDistinct(asList("C#", "C++"));
+		assertEquals(found.size(), 2);
+	}
```
Comment (leocwolter) [inline src/test/java/org/mamute/dao/TagDAOTest.java]: the test `should_not_repeat_tags_with_similar_name` already covers that, no?

File: src/test/java/org/mamute/dao/TagDAOTest.java [hunk 17]
```diff
@@ -88,17 +96,32 @@ public void should_get_main_tags_of_the_provided_user() throws Exception {
...
-		assertEquals(2, tagsNames.size());
-		assertEquals(java.getName(), tagsNames.get(0));
-		assertEquals(ruby.getName(), tagsNames.get(1));
-	}
+		assertEquals(5, tagsNames.size());
+		assertEquals(csharp.getName(), tagsNames.get(0));
+		assertEquals(cplus.getName(), tagsNames.get(1));
+		assertEquals(weirdCharacters.getName(), tagsNames.get(2));
+        assertEquals(java.getName(), tagsNames.get(3));
+        assertEquals(ruby.getName(), tagsNames.get(4));
```
Comment (leocwolter) [inline src/test/java/org/mamute/dao/TagDAOTest.java]: you could use hamcrest matchers to make it easier to read here. Something like:
```
assertThat(tagNames, Matchers.containsInAnyOrder(java.getName(), ruby.getName()....)
```

File: src/main/java/org/mamute/controllers/TagPageController.java (hunk 9 above)
Comment (MatejBalantic) [inline src/main/java/org/mamute/controllers/TagPageController.java]: I can change this if you will insist. I personally prefer keeping the structure of the relations on the model as-is, don't like to flatten it, as it's harder to understand the relationships there. I mean, that why the models are there, right :) But if you feel strongly about it, I can change it

File: src/main/java/org/mamute/infra/NormalizerBrutal.java (hunk 11 above)
Comment (MatejBalantic) [inline src/main/java/org/mamute/infra/NormalizerBrutal.java]: You are right that I didn't do the best job explaining what is happening.

So this is what we want:
- normalize characters like čćšž etc. (also called base-glyph, they match with P{M} in unicode)
- NOT normalize characters like  + or # or ! or . and rather URL encode them

If we would normalize character like #, it would first be decomposed to several characters and than URL encoded, which would result in something that is in fact not URL-encoded representation of a hash. 

In case we don't want to URL encode, then we SHOULD normalize all characters and strip out non ASCII representation of them at the end (thus \w)

Comment (leocwolter) [inline src/main/java/org/mamute/infra/NormalizerBrutal.java]: Oh, okay, it makes sense, sorry :)

File: src/main/java/org/mamute/controllers/TagPageController.java (hunk 9 above)
Comment (leocwolter) [inline src/main/java/org/mamute/controllers/TagPageController.java]: Well, the problem is if `getTag` starts to receive an argument at some point, we would need to pass this new argumento at every place we call `tagPage.getTag("argument").getUriName()`.
But I understand your point, in that case we could just use a variable to keep this value so we don't need to call it multiple times :)

Comment (MatejBalantic) [inline src/main/java/org/mamute/controllers/TagPageController.java]: Since this are DOMs getTag will probably never receive any arguments - it's a to-one relationship, so what could the argument be? It also feels as over-designing, but I'll do it since I don't feel so strongly about it :)

File: src/main/java/org/mamute/infra/NormalizerBrutal.java (hunk 12 above)
Comment (MatejBalantic) [inline src/main/java/org/mamute/infra/NormalizerBrutal.java]: will do

File: src/main/java/org/mamute/dao/TagDAO.java (hunk 10 above)
Comment (MatejBalantic) [inline src/main/java/org/mamute/dao/TagDAO.java]: will do

File: src/test/java/org/mamute/dao/TagDAOTest.java (hunk 15 above)
Comment (MatejBalantic) [inline src/test/java/org/mamute/dao/TagDAOTest.java]: Good catch

File: src/test/java/org/mamute/dao/TagDAOTest.java (hunk 16 above)
Comment (MatejBalantic) [inline src/test/java/org/mamute/dao/TagDAOTest.java]: Good catch!

File: src/test/java/org/mamute/dao/TagDAOTest.java (hunk 17 above)
Comment (MatejBalantic) [inline src/test/java/org/mamute/dao/TagDAOTest.java]: Thanks, done

**Comment by MatejBalantic** [comment]
I've addressed the review comments, please review when you can

File: src/test/java/org/mamute/infra/NormalizerBrutalTest.java [hunk 18]
```diff
@@ -27,6 +32,9 @@ public void should_remove_question_marks_and_bangs() {
 		String text = "como faz para normalizar uma string?!";
 		String slug = NormalizerBrutal.toSlug(text);
 		assertEquals("como-faz-para-normalizar-uma-string", slug);
+
+		String slugEncoded = NormalizerBrutal.toSlug(text, true);
+		assertEquals("como-faz-para-normalizar-uma-string%3f%21", slugEncoded);
```
Comment (csokol) [inline src/test/java/org/mamute/infra/NormalizerBrutalTest.java]: If we start url encoding like this, we will lose the friendly urls we have for questions (example: http://demo.mamute.org/297-what-happens-if-i-use-strange-chars-on-the-title). 

Maybe we should keep the old normaliser and create a normaliser only for tags. What do you think, @MatejBalantic?

Comment (JamesSullivan) [inline src/test/java/org/mamute/infra/NormalizerBrutalTest.java]: Apologies for jumping in. The current pull request is not usable with Asian languages. I would think just using URL encoding is the simplest and covers the most cases for the least work particularly as browsers now [handle the display of Unicode urls](http://stackoverflow.com/questions/2742852/unicode-characters-in-urls) better. There are a few characters such as "?" that are still displayed as % hex code in the url input box but there are good security reasons for that. Realizing this is a question site and most titles will end in a question mark that would show up as %3F  perhaps the only time a character should be stripped is when the title ends in a question mark.

Comment (JamesSullivan) [inline src/test/java/org/mamute/infra/NormalizerBrutalTest.java]: We can support non-latin languages by changing one line of code and adding another in NormalizerBrutal. 
[NormalizerBrutal.java.txt](https://github.com/caelum/mamute/files/95251/NormalizerBrutal.java.txt)

All tests pass except for this one

**Comment by felipeweb** [comment]
closing here see https://github.com/caelum/mamute/issues/119#issuecomment-192878835 for more details

**Comment by JamesSullivan** [comment]
This fix (disallowing most characters) although simpler does not allow Japanese or Chinese tags, etc. The updated brutalizer and tag slugging would have allowed any Unicode language.
